File: schema/src/main/java/dev/sayaya/handbook/domain/Attribute.java
package dev.sayaya.handbook.domain;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonProperty;
import jsinterop.annotations.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;

@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
@JsType(isNative = true, namespace = JsPackage.GLOBAL, name = "Object")
@Getter(onMethod_ = {@JsOverlay, @JsIgnore})
@Setter(onMethod_ = {@JsOverlay, @JsIgnore})
@Accessors(fluent = true)
@NoArgsConstructor
public final class Attribute {
    @JsonProperty("id") @JsProperty(name = "id") private String id;
    @JsonProperty("name") @JsProperty(name = "name") private String name;
    @JsonProperty("order") @JsProperty(name = "order") private int order;
    @JsonProperty("type") @JsProperty(name = "type") private AttributeType type;
    @JsonProperty("description") @JsProperty(name = "description") private String description;
    @JsonProperty("nullable") @JsProperty(name = "nullable") private boolean nullable;
    @JsonProperty("inherited") @JsProperty(name = "inherited") private boolean inherited;
    @JsonProperty("read_roles") @JsProperty(name = "read_roles") private String[] readRoles;
    @JsonProperty("write_roles") @JsProperty(name = "write_roles") private String[] writeRoles;

    @JsOverlay @JsIgnore
    public static Attribute create(String id, String name, int order, AttributeType type) {
        if (name == null || name.trim().isEmpty()) throw new IllegalArgumentException("Attribute name cannot be blank.");
        if (!name.matches("^[a-zA-Z0-9_-]+$")) throw new IllegalArgumentException("Attribute name can only contain alphanumerics, hyphens, and underscores.");
        Attribute attr = new Attribute();
        attr.id(id);
        attr.name(name);
        attr.order(order);
        attr.type(type);
        return attr;
    }

    @JsOverlay @JsIgnore
    public Attribute withName(String name) {
        Attribute attr = create(id(), name, order(), type());
        attr.description(description());
        attr.nullable(nullable());
        attr.inherited(inherited());
        attr.readRoles(readRoles());
        attr.writeRoles(writeRoles());
        return attr;
    }

    @JsOverlay @JsIgnore
    public Attribute withType(AttributeType type) {
        Attribute attr = create(id(), name(), order(), type);
        attr.description(description());
        attr.nullable(nullable());
        attr.inherited(inherited());
        attr.readRoles(readRoles());
        attr.writeRoles(writeRoles());
        return attr;
    }

    @JsOverlay @JsIgnore
    public Attribute withDescription(String description) {
        Attribute attr = create(id(), name(), order(), type());
        attr.description(description);
        attr.nullable(nullable());
        attr.inherited(inherited());
        attr.readRoles(readRoles());
        attr.writeRoles(writeRoles());
        return attr;
    }

    @JsOverlay @JsIgnore
    public Attribute cloneWithoutId() {
        Attribute attr = create(null, name(), order(), type());
        attr.description(description());
        attr.nullable(nullable());
        attr.inherited(inherited());
        attr.readRoles(readRoles());
        attr.writeRoles(writeRoles());
        return attr;
    }
}

File: schema/src/main/kotlin/dev/sayaya/handbook/domain/Attribute.kt
package dev.sayaya.handbook.domain

import java.io.Serializable

/**
 * 'Type'을 구성하는 개별 속성을 정의하는 값 객체(Value Object).
 *
 * 이 객체는 고유한 ID를 갖지 않으며, 모든 프로퍼티의 값들이 모여 하나의 '속성'을 정의합니다.
 *
 * @property name 속성의 이름 (e.g., "title", "author")
 * @property order GUI 등에서 표시될 순서
 * @property description 속성에 대한 설명
 * @property type 이 속성의 데이터 타입 (e.g., Text, Number, Document 등)
 * @property nullable 이 속성의 값이 null을 허용하는지 여부
 * @property inherited 이 속성이 부모 'Type'으로부터 상속되었는지 여부
 */
data class Attribute (
    val name: String,
    val order: Short,
    val description: String?,
    val type: AttributeType,
    val nullable: Boolean,
    val inherited: Boolean
) : Serializable {
    init {
        require(name.isNotBlank()) { "Attribute name cannot be blank." }
        require(name.matches(Regex("^[a-zA-Z0-9_-]+$"))) { "Attribute name can only contain alphanumerics, hyphens, and underscores." }
    }
}